token/ts/src/instructions/approve.ts

```diff
@@ -50,3 +60,44 @@ export function createApproveInstruction(
 
     return new TransactionInstruction({ keys, programId, data });
 }
+
+/** TODO: docs */
+export interface DecodedApproveInstruction {
+    instruction: TokenInstruction.Approve;
+    account: AccountMeta;
+    delegate: AccountMeta;
+    owner: AccountMeta;
+    multiSigners: AccountMeta[];
+    amount: bigint;
+}
+
+/**
+ * Decode an Approve instruction
+ *
+ * @param instruction Transaction instruction to decode
+ * @param programId   SPL Token program account
+ *
+ * @return Decoded instruction
+ */
+export function decodeApproveInstruction(
+    instruction: TransactionInstruction,
+    programId = TOKEN_PROGRAM_ID
+): DecodedApproveInstruction {
+    if (!instruction.programId.equals(programId)) throw new TokenInvalidInstructionProgramError();
+
+    const [account, delegate, owner, ...multiSigners] = instruction.keys;
+    if (!account || !delegate || !owner) throw new TokenInvalidInstructionKeysError();
+
+    if (instruction.data.length !== approveInstructionData.span) throw new TokenInvalidInstructionTypeError();
+    const data = approveInstructionData.decode(instruction.data);
+    if (data.instruction !== TokenInstruction.Approve) throw new TokenInvalidInstructionDataError();
+
+    return {
+        instruction: data.instruction,
+        account,
+        delegate,
+        owner,
+        multiSigners,
+        amount: data.amount,
+    };
+}
```

Comment:
@joncinque can I get your thoughts on this? Decoding token instructions to validate them is a need for a few things I'm building, but I want to see if there's a better way to think about doing this.

I'm especially wondering if we should validate the existence account keys at all, and if we should try validate them to be correct in the client side (for example, checking for isWritable and isSigner, and checking for multiSigners if the owner/authority is not a signer).

An argument for is that it would allow someone to generally verify whether an instruction was likely to succeed or fail before simulating it, and it lets us return stricter types from the decode functions. An argument against is that an instruction missing or having the wrong keys is still a valid type of that instruction, even if it will fail.

Reply:
Since I'm not sure where these decodes are ever used, you can define the use-case entirely.  For example, the explorer decodes things in its own way, without relying on any of these decoders.

Correct me if I'm wrong, but it seems like there are two main use-cases:
* validating the instruction is correct, likely before sending
* decoding a sent instruction fetched from somewhere else, which may be incorrect

In that sense, perhaps we should have two of these, `checked` and `unchecked`, where the `unchecked` just does the plain serde, and `checked` actually makes sure that the instruction is well-formed, with correct writable / signer flags and numbers of accounts.  The default is the `checked` version.  What do you think?

Reply:
Yeah, I think you're right about the use cases. That's a good idea, I'll do that.